src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -359,6 +385,379 @@ def to_Spec {f : A} (m : ℕ) (f_deg : f ∈ 𝒜 m) :
 
 end
 
+namespace from_Spec
+
+open graded_algebra set_like finset (hiding mk_zero)
+
+variables {𝒜} {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m}
+
+private meta def mem_tac : tactic unit :=
+let b : tactic unit :=
+  `[exact pow_mem_graded _ (submodule.coe_mem _) <|> exact nat_cast_mem_graded _ _] in
+b <|> `[by repeat { all_goals { apply graded_monoid.mul_mem } }; b]
+
+/--The function from `Spec A⁰_f` to `Proj|D(f)` is defined by `q ↦ {a | aᵢᵐ ∈ q}`, i.e. send `q` a
+prime ideal in `A⁰_f` to the homogeneous prime relevant ideal containing only and all the element
+`a : A` such that the `m`-th power of `i`-th projection of `a` is in `q`.
+
+The set `{a | aᵢᵐ ∈ q}`
+* is an ideal is proved in  `carrier.as_ideal`;
+* is homogeneous is proved in `carrier.as_homogeneous_ideal`;
+* is prime is proved in `carrier.as_ideal.prime`;
+* is relevant is proved in `carrier.as_ideal.relevant`
+-/
+def carrier (q : Spec.T (A⁰_ f_deg)) : set A :=
+{a | ∀ i, (⟨mk ((proj 𝒜 i a)^m) ⟨_, ⟨_, rfl⟩⟩, ⟨i, ⟨_, by mem_tac⟩, rfl⟩⟩ : A⁰_ f_deg) ∈ q.1 }
+
+lemma mem_carrier_iff (q : Spec.T (A⁰_ f_deg)) (a : A) :
+  a ∈ carrier q ↔
+  ∀ i, (⟨mk ((proj 𝒜 i a)^m) ⟨_, ⟨_, rfl⟩⟩, ⟨i, ⟨_, by mem_tac⟩, rfl⟩⟩ : A⁰_ f_deg) ∈ q.1 :=
+iff.rfl
+
+lemma carrier.zero_mem (hm : 0 < m) (q : Spec.T (A⁰_ f_deg)) :
+  (0 : A) ∈ carrier q := λ i,
+by simpa only [linear_map.map_zero, zero_pow hm, mk_zero] using submodule.zero_mem _
+
+lemma carrier.add_mem (q : Spec.T (A⁰_ f_deg)) {a b : A} (ha : a ∈ carrier q) (hb : b ∈ carrier q) :
+  a + b ∈ carrier q :=
```

Comment:
You may want to look into [the proof that `ideal.radical` is closed under addition](
https://github.com/leanprover-community/mathlib/blob/63138639ca195344ae96aa77f3a02b90a3ac5c68/src/ring_theory/ideal/operations.lean#L599) and see if you can adapt it to golf this.

Reply:
I made `carrier.add_mem` shorter but not as succinct as `ideal.radical`'s add_mem. I will do similar things for `smul_mem` etc as well

Reply:
The rest of the proofs are golfed as well

Reply:
I've golfed the proof down to 25 lines and halved the compile time. Maybe it can inspire you to golf the other proofs, but if not I'll try them later.
```
lemma carrier.add_mem (q : Spec.T (A⁰_ f_deg)) {a b : A} (ha : a ∈ carrier q) (hb : b ∈ carrier q) :
  a + b ∈ carrier q :=
begin
  refine λ i, (q.2.mem_or_mem _).elim id id,
  change subtype.mk (localization.mk _ _ * mk _ _) _ ∈ q.1,
  simp_rw [mk_mul, ← pow_add, map_add, add_pow, mk_sum, mul_comm, ← nsmul_eq_mul, ← smul_mk],
  let g : ℕ → A⁰_ f_deg := λ j, (m + m).choose j • if h2 : m + m < j then 0 else if h1 : j ≤ m
    then ⟨mk (proj 𝒜 i a ^ j * proj 𝒜 i b ^ (m - j)) ⟨_, i, rfl⟩, i, ⟨_, _⟩, rfl⟩ *
      ⟨mk (proj 𝒜 i b ^ m) ⟨_, i, rfl⟩, i, ⟨_, by mem_tac⟩, rfl⟩
    else ⟨mk (proj 𝒜 i a ^ m) ⟨_, i, rfl⟩, i, ⟨_, by mem_tac⟩, rfl⟩ *
      ⟨mk (proj 𝒜 i a ^ (j - m) * proj 𝒜 i b ^ (m + m - j)) ⟨_, i, rfl⟩, i, ⟨_, _⟩, rfl⟩,
  rotate,
  { rw (_ : m * i = _), mem_tac, rw [← add_smul, nat.add_sub_of_le h1], refl },
  { rw (_ : m * i = _), mem_tac, rw ← add_smul, congr,
    zify [le_of_not_lt h2, le_of_not_le h1], abel },
  convert_to ∑ i in range (m + m + 1), g i ∈ q.1, swap,
  { refine q.1.sum_mem (λ j hj, nsmul_mem _ _), split_ifs,
    exacts [q.1.zero_mem, q.1.mul_mem_left _ (hb i), q.1.mul_mem_right _ (ha i)] },
  apply subtype.ext,
  rw [degree_zero_part.coe_sum, subtype.coe_mk],
  apply finset.sum_congr rfl (λ j hj, _),
  congr' 1, split_ifs with h2 h1,
  { exact ((finset.mem_range.1 hj).not_le h2).elim },
  all_goals { simp only [subtype.val_eq_coe, degree_zero_part.coe_mul, subtype.coe_mk, mk_mul] },
  { rw [mul_assoc, ← pow_add, add_comm (m - j), nat.add_sub_assoc h1] },
  { rw [← mul_assoc, ← pow_add, nat.add_sub_of_le (le_of_not_le h1)] },
end
```